File: 0076-minimum-window-substring/0076-minimum-window-substring.cpp
class Solution {
public:
    string minWindow(string s, string t) {
        if (t == "")
            return "";
        unordered_map<char, int> t_map;
        unordered_map<char, int> s_map;
        for (char c : t) {
            t_map[c]++;
        }

        int start = 0, end = 0, have = 0, need = t_map.size(), minDis = INT_MAX,
            left = 0, right = 0;
        while (end < s.size()) {

            if (t_map.find(s[end]) != t_map.end()) {
                s_map[s[end]]++;
                if (s_map[s[end]] == t_map[s[end]]) {
                    have++;
                }
            }

            while (need == have) {
                if (t_map.find(s[start]) != t_map.end()) {
                    s_map[s[start]]--;
                    if (s_map[s[start]] < t_map[s[start]]) {
                        have--;
                    }
                }
                if (end - start + 1 < minDis) {
                    minDis = end - start + 1;
                    left = start;
                }
                start++;
            }
            end++;
        }

        if (minDis == INT_MAX) {
            return "";
        }

        return s.substr(left, minDis);
    }

};

File: 0076-minimum-window-substring/0076-minimum-window-substring.py
class Solution:
    def minWindow(self, s: str, t: str) -> str:
        if t == "": return ""

        win, target = {}, {}

        for n in t:
            target[n] = 1 + target.get(n, 0)    # get the value of n if n exists in the map(target) or give default value (0)
        
        have, need = 0, len(target) 
        left, right = 0, 0
        res, reslen = "", float("inf")

        for right in range(len(s)):
            char = s[right]
            if char in target:
                win[char] = 1 + win.get(char, 0)
            
                if win[char] == target[char]:
                    have += 1

            while have == need:
                # update our result
                if right - left + 1 < reslen:
                    res = s[left:right+1]
                    reslen = right - left + 1

                # pop from the left of our window
                if s[left] in target:
                    win[s[left]] -= 1
                    # update the have variable
                    if win[s[left]] < target[s[left]]:
                        have -= 1

                left += 1
        return res